basic/code17.s

```diff
@@ -4,36 +4,45 @@ conint	jsr posint
 	bne gofuc
 	ldx faclo
 	jmp chrgot
-val	jsr len1
-val_str	bne @1
+
+; the "val" function takes a string and turns it into a number by interpreting
+; the ascii digits etc. Except for the problem that a terminator must be
+; supplied by replacing the character beyond the string, VAL is merely a call
+; to floating point input ("finh").
+val	jsr len1		;get length
+val_str	bne val1	;return 0 if len=0
 	jmp zerofc
-@1:	ldx txtptr
+val1	ldx txtptr
 	ldy txtptr+1
 	stx strng2
 	sty strng2+1
 	ldx index1
 	stx txtptr
+	lda index1		;load the low byte of index1
 	clc
-	adc index1
-	sta index2
+	adc len1		;add len to index1 and put in strng2
+	sta strng2
+	lda index1+1	;load the high byte of index1
+	adc #0
+	sta strng2+1	;store result in high byte of strng2
 	ldx index1+1
 	stx txtptr+1
 	bcc val2
 	inx
-val2	stx index2+1
-	lda (index2)
+val2	stx strng2+1
+	lda (strng2),y	;replace character after string with $00 (fake eol)
```

Comment:
What is the value of Y here?  It's the high byte of txtptr, which makes no sense as an offset to `strng2`.

I don't think this PR solves the issue.  In fact, it breaks `VAL()` when you give it a proper argument.

Reply:
![image](https://github.com/user-attachments/assets/79902443-82a1-434b-a54d-598d370e279f)

Reply:
![image](https://github.com/user-attachments/assets/b3c5b16b-7648-4138-90ce-f82fc5f87f42)

Reply:
Intresting, because I used `STA (),Y` and it seemed to fix the issue, but now that you told me that it breaks `VAL()` when given real arguments, I have to re-work my solution.